metaseq/data/document_to_sequence.py

```diff
@@ -0,0 +1,377 @@
+
+        baserng = np.random.default_rng(self.shuffle_buffer_seed)
+        rngstate = None
+        rngcount = 0
+
+        def create_rng():
+            nonlocal rngstate, rngcount
+            while True:
+                rngstate = baserng._bit_generator.__getstate__()
+                # overhead to calling 'integers' is high,
+                # so request random numbers in batches of 1024
+                for x in baserng.integers(self.shuffle_buffer_size, size=1024):
+                    rngcount += 1
+                    yield x
+
+        rng = None if self.shuffle_buffer_seed is None else iter(create_rng())
+
+        buffer = []
+
+        def get_next_item_and_replace_in_buffer(replacement_item: Sequence) -> Sequence:
+            # return a random item from the buffer and replace with a new item
+            if rng is not None:
+                if len(buffer) == self.shuffle_buffer_size:
+                    idx = next(rng)
+                else:
+                    if len(buffer) == self.shuffle_buffer_size - 1:
+                        # when the buffer drains at the end we are asking
+                        # for a smaller range of random numbers than we have batched.
+                        # To match our previous behavior, reset the
+                        # state of base rng to what it would have been
+                        # in previous version of the code to generate
+                        # these smaller numbers (only happens at the very end)
+                        if rngstate is not None:
+                            baserng._bit_generator.__setstate__(rngstate)
+                        baserng.integers(self.shuffle_buffer_size, size=rngcount % 1024)
+                    idx = baserng.integers(len(buffer))
+            else:
+                idx = 0
+            item = buffer[idx]
+            if replacement_item is not None:
+                buffer[idx] = replacement_item
+            else:
+                buffer.pop(idx)
+            return item
+
+        def sequences() -> Iterable[Sequence]:
+            for block in block_itr:
+                if len(buffer) < self.shuffle_buffer_size:
+                    # initially fill the buffer to the requested size
+                    buffer.append(block)
+                else:
+                    # return random block from the buffer and replace with new block
+                    yield get_next_item_and_replace_in_buffer(block)
+
+            # clear buffer of any remaining items
+            while buffer:
+                yield get_next_item_and_replace_in_buffer(None)
+
+        for i, elem in enumerate(sequences()):
+            if i >= to_skip:
+                elem["ids"] = torch.LongTensor(elem["ids"])
+                subsequences = []
+                for c, start, ln in elem["block"]:
+                    # A padding tensor (<padding_value>, 0, length)
+                    if c[0] == "padding":
+                        subsequences.append(
```

Comment:
Just to see if I follow from looking at type signatures: `elem` here is a `Sequence`, which means `elem["block"]` is a `List[SequenceFragment]`; the for loop here is then iterating through `c, start, ln` of `SequenceFragment`s, which means `c` is a `List[Document]`, which means `c[0]` is a `Document`?  

There's some magic with how `"padding"` state is being passed through possible block iterators (i.e. the `yield_*` methods?) so just wanting to confirm logic around this bit since `Document` annotation above now seems to be Union of `Tuple[...], str`..?

Reply:
I updated the documentation in the code to hopefully document this better, and document the rest of the file better.  Documents are a Union of int (index into the self.dataset of unloaded document, the string "padding" representing a document full of self.padding_idx, and torch.Tensor (already tokenized data). The Tuple thing was a bug in the annotations: the document iterator returns a tuple of `(size, [document])`. "padding" is the way the block iterators communicate that part of the sequence should be filled will padding. It is easiest to think of it conceptually as a tensor full of padding_idx of infinite size.

Please let me know what remains as the trickiest stuff to understand. I want to make sure this code stay maintainable while also being fast enough to do fast-forwarding without much noticeable cost.